### PR TITLE
fix(indexed-db): reset now clears db tables

### PR DESCRIPTION
**Pull request by fczuardi**
Co-authored-by: @taboca
Co-authored-by: @mquasar

### Describe the changes you have made in this PR

_A clear and concise description of what you want to happen_

### Link this PR to an issue [optional]

Fix: #2435

### Type of change

(Remove other not matching type)

- `fix`: Bug fix (non-breaking change which fixes an issue)
- `feat`: New feature (non-breaking change which adds functionality)
- `feat!`: Breaking change (fix or feature that would cause existing functionality to not work as expected)
- `docs`: Documentation update

### Screenshots of the changes [optional]

_Add screenshots to make your changes easier to understand. You can also add a video here._

### How has this been tested?

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration_

### Checklist

- [ ] My code follows the style guidelines of this project and performed a self-review of my own code
- [ ] New and existing tests pass locally with my changes
- [ ] I checked if I need to make corresponding changes to the documentation (and made those changes if needed)

### CODE REVIEW

File: src/extension/background-script/db.ts [hunk 1]
```diff
@@ -86,6 +86,21 @@ export class DB extends Dexie {
     return true;
   }
 
+  async clearAllTables() {
+    await this.allowances.clear();
+    await this.payments.clear();
+    await this.blocklist.clear();
+    await this.permissions.clear();
+
+    await browser.storage.local.set({
```
Comment (rolznz) [inline src/extension/background-script/db.ts:95]: There is also browser.storage.local.clear() - would this be better to ensure future entries are also cleared?

**Review by reneaaron — CHANGES_REQUESTED** [review]
Thanks for the PR! 

As @rolznz noted I think the `browser.storage.clear()` APIs would be a better fit here as they delete all data and do not need to be adjusted in case there are new tables added in future.

File: src/extension/background-script/db.ts (hunk 1 above)
Comment (mquasar) [reply src/extension/background-script/db.ts:95]: From what I've figured from the code, there's a sort of replication/mirroring of browser.storage.local and IndexedDB. Also, 'to keep consistency' with the code.

**Comment by mquasar** [comment]
> Thanks for the PR! 
> 
> As @rolznz noted I think the `browser.storage.clear()` APIs would be a better fit here as they delete all data and do not need to be adjusted in case there are new tables added in future. 

We tried that but the data still 'lives on' in the IndexedDB. I might be wrong here, obviously 😊

**Comment by taboca** [comment]
> Thanks for the PR!
> 
> As @rolznz noted I think the `browser.storage.clear()` APIs would be a better fit here as they delete all data and do not need to be adjusted in case there are new tables added in future.

Although there is a situation that loading data ( such as payments ) comes from browser.storage API - which would indeed suggest to that deleting browser.storage is enough - yet based on local and limited tests, we have stumbled in a situation where IndexedDB data was persistent. 

A part of the code that gives us a hint the function loadFromStorage: 

https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#L94

In addition, notice how the code logic gives up the sync fro the browser.storage API in the situation where IndexedDB data is already in there: 

https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#LL116C6-L129C12

File: src/extension/background-script/db.ts (hunk 1 above)
Comment (taboca) [reply src/extension/background-script/db.ts:95]: I also felt that the notion of "attribute by attribute" clearing would be more verbose - at this point - in contrast with the idea of a whole wipe using browser.storage.local.clear() and the whole wipe in the DB too.

**Comment by rolznz** [comment]
> > Thanks for the PR!
> > As @rolznz noted I think the `browser.storage.clear()` APIs would be a better fit here as they delete all data and do not need to be adjusted in case there are new tables added in future.
> 
> Thanks for supporting us @reneaaron and @rolznz
> 
> Although there is a situation that loading data ( such as payments ) comes from browser.storage API - which would indeed suggest that just deleting browser.storage is enough - yet based on local and limited tests, we have stumbled in a situation where IndexedDB data was persistent.
> 
> The part of the code that gives us that idea is the function loadFromStorage:
> 
> https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#L94
> 
> In addition, notice the following section example where the logic gives up the load sync from browser.storage API in the situation where IndexedDB data exists: https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#LL116C6-L129C12
> 
> With the above said it is correct that there is a situation where IndexedDB is not present, which indeed require us to also clear the data from browser.storage
> 
> https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#L106
> 
> Some other observations [#2435 (comment)](https://github.com/getAlby/lightning-browser-extension/issues/2435#issuecomment-1559747137)
> 
> With the above said, if the suggestion has to do with easy maintenance, and if we do the "browser.storage.clear()" then we would have to keep some consistency with wiping the IndexedDB database entirely too? Another idea is to try to do the maintenance from the loadFromStorage function in db.ts which is the code that gets kicked in app initialization time.

@mquasar @taboca thanks for testing and looking deeper into the issue. I think the goal is to delete absolutely everything when resetting the extension. So any and all storage mechanisms (indexedDb, browser.storage, etc) we save data to should be completely cleared.

**Comment by taboca** [comment]
> > > Thanks for the PR!
> > > As @rolznz noted I think the `browser.storage.clear()` APIs would be a better fit here as they delete all data and do not need to be adjusted in case there are new tables added in future.
> > 
> > 
> > Thanks for supporting us @reneaaron and @rolznz
> > Although there is a situation that loading data ( such as payments ) comes from browser.storage API - which would indeed suggest that just deleting browser.storage is enough - yet based on local and limited tests, we have stumbled in a situation where IndexedDB data was persistent.
> > The part of the code that gives us that idea is the function loadFromStorage:
> > https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#L94
> > 
> > In addition, notice the following section example where the logic gives up the load sync from browser.storage API in the situation where IndexedDB data exists: https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#LL116C6-L129C12
> > With the above said it is correct that there is a situation where IndexedDB is not present, which indeed require us to also clear the data from browser.storage
> > https://github.com/getAlby/lightning-browser-extension/blob/5bddad27cd0f58dcb83ffbceb5f3cf1317bb517c/src/extension/background-script/db.ts#L106
> > 
> > Some other observations [#2435 (comment)](https://github.com/getAlby/lightning-browser-extension/issues/2435#issuecomment-1559747137)
> > With the above said, if the suggestion has to do with easy maintenance, and if we do the "browser.storage.clear()" then we would have to keep some consistency with wiping the IndexedDB database entirely too? Another idea is to try to do the maintenance from the loadFromStorage function in db.ts which is the code that gets kicked in app initialization time.
> 
> @mquasar @taboca thanks for testing and looking deeper into the issue. I think the goal is to delete absolutely everything when resetting the extension. So any and all storage mechanisms (indexedDb, browser.storage, etc) we save data to should be completely cleared.

Ok, I will check then the other aspects of DB that seem to be there, such as "LBE-AVAILABILITY-CHECK" and the ways to clear all the DBs.

**Comment by fczuardi** [comment]
> Thanks for the PR!
> 
> As @rolznz noted I think the `browser.storage.clear()` APIs would be a better fit here as they delete all data and do not need to be adjusted in case there are new tables added in future.

We have just updated the patch with the suggestions. It's much simpler now :) Thanks @rolznz !!

**Comment by rolznz** [comment]
@fczuardi I made some more simplifications. Could you please review the changes?

File: src/extension/background-script/actions/setup/reset.ts [hunk 2]
```diff
@@ -1,16 +1,11 @@
...
 const reset = async (message: MessageReset) => {
-  await state.getState().password(null);
-  state.setState({
-    accounts: {},
-    account: null,
-    connector: null,
-    currentAccountId: null,
-  });
-  await state.getState().saveToStorage();
+  await state.getState().reset();
```
Comment (fczuardi) [inline src/extension/background-script/actions/setup/reset.ts:7]: I see that `reset()` will change back `accounts` and `currentAccountId` back to the default values, but I couldn't see if `account` and `connector` are being reset as well. Should they?

**Comment by fczuardi** [comment]
> @fczuardi I made some more simplifications. Could you please review the changes?

@rolznz from your 3 commits, the first 2 LGTM, the third is probably fine as well, I just left a comment asking if your version is equivalent because I couldn't see 2 attributes nullified.

File: src/extension/background-script/actions/setup/reset.ts (hunk 2 above)
Comment (rolznz) [reply src/extension/background-script/actions/setup/reset.ts:7]: @fczuardi thanks for the review! I think I have fixed the issue now.

**Comment by rolznz** [comment]
@reneaaron I think this PR is ready for review now. 💪

**Comment by bumi** [comment]
looks good to me. @reneaaron have a look and merge.

**Review by reneaaron — APPROVED** [review]
Thanks for the PR @fczuardi! :muscle: 

tACK

**Pull request merged**